sdks/python/apache_beam/ml/gcp/visionml.py

```diff
@@ -0,0 +1,526 @@
+            features=self.features,
+            output_config=self.output_config,
+            retry=self.retry,
+            timeout=self.timeout,
+            batch_size=self.batch_size,
+            client_options=self.client_options,
+            metadata=self.metadata))
+
+
+@typehints.with_input_types(
+    Union[text_type, binary_type], Optional[vision.types.ImageContext])
+class _AsyncBatchImageAnnotateFn(DoFn):
+  """A DoFn that sends each input element to the GCP Vision API
+  service in batches and stores the results in GCS. Returns a
+  ``google.cloud.vision.types.AsyncBatchAnnotateImagesResponse`` containing the
+  the output location and metadata from AsyncBatchAnnotateImagesRequest.
+  """
+  def __init__(
+      self,
+      features,
+      output_config,
+      retry,
+      timeout,
+      client_options,
+      batch_size,
+      metadata):
+
+    super(_AsyncBatchImageAnnotateFn, self).__init__()
+    self._client = None
+    self.features = features
+    self.output_config = output_config
+    self.retry = retry
+    self.timeout = timeout
+    self.client_options = client_options
+    self.batch_size = batch_size
+    self._batch_elements = None
+    self.metadata = metadata
+    self.counter = Metrics.counter(self.__class__, "API Calls")
+
+  def start_bundle(self):
+    self._client = get_vision_client(self.client_options)
+    self._batch_elements = []
+
+  def finish_bundle(self):
+    if self._batch_elements:
+      response = self._flush_batch()
+      self.counter.inc()
+      return response
+
+  def create_annotate_image_request(self, element, image_context):
+    if isinstance(element, text_type):
+      image = vision.types.Image(
+          source=vision.types.ImageSource(image_uri=element))
+    else:  # Typehint checks only allows text_type or binary_type
+      image = vision.types.Image(content=element)
+
+    request = vision.types.AnnotateImageRequest(
+        image=image, features=self.features, image_context=image_context)
+    return request
+
+  def process(self, element, context_side_input=None, *args, **kwargs):
+    if context_side_input:  # If we have a side input image context, use that
+      image_context = context_side_input.get(element)
+    else:
+      image_context = None
+
+    # Evaluate batches
+    request = self.create_annotate_image_request(element, image_context)
+    self._batch_elements.append(request)
+    if len(self._batch_elements) >= self.batch_size or \
+            len(self._batch_elements) >= AsyncBatchAnnotateImage.MAX_BATCH_SIZE:
+      response = self._flush_batch()
+      self.counter.inc()
+      yield response
+
+  def _flush_batch(self):
+    operation = self._client.async_batch_annotate_images(
+        requests=self._batch_elements,
+        output_config=self.output_config,
+        retry=self.retry,
+        timeout=self.timeout,
+        metadata=self.metadata)
+    self._batch_elements = []
+    return operation
```

Comment:
On closer examination, it may be that we need to wait for the return result of the operation anyway, like so: `return operation.results()` as without it, the function returns a `google.api_core.operation.Operation object` and then Beam runs into the following error:
```
AttributeError: Can't pickle local object 'if_exception_type.<locals>.if_exception_type_predicate' [while running 'Annotate image/ParDo(_AsyncBatchImageAnnotateFn)']
```

Reply:
That is, if this PTransform should return anything at all.

Perhaps it makes more sense to not return anything and just send off the Vision API call. The "downside" is that this PTransform (AsyncBatchAnnotateImage) would essentially become a sink).

Reply:
I've answered to this in the main thread.